### PR TITLE
chore: removes eslint warnings from item simplification

### PR DESCRIPTION
**Pull request by jonaslagoni**
**Description**
Refactors the item simplification to remove eslint warnings.

- This reduced the number of calls to merge since we make more checks in regards to when to call the function.
- Adapted tests to always return undefined because we mock functions

**Related issue(s)**
related to #53

### CODE REVIEW

**Review by magicmatatjahu — DISMISSED** [review]

**Comment by jonaslagoni** [comment]
So the error occurred with circular schemas being returned twice. So I added a check and removed the model if it already existed (since it will be re-added right)

File: src/simplification/SimplifyItems.ts [hunk 1]
```diff
@@ -12,50 +13,66 @@ type Output = CommonModel | undefined;
...
+ * Go through schema(s) and combine the simplified items together.
+ * 
+ * @param schema to go through
+ * @param currentModel the current output
+ * @param simplifier the simplifier to use
+ * @param seenSchemas schemas which we already have outputs for
+ */
+function combineSchemas(schema: (Schema | boolean) | (Schema | boolean)[] | undefined, currentModel: Output, simplifier : Simplifier, seenSchemas: Map<any, Output>) {
+  if (schema === undefined) return;
+  if (typeof schema === 'boolean') return;
```
Comment (magicmatatjahu) [inline src/simplification/SimplifyItems.ts:56]:
```suggestion
  if (typeof schema !== 'object') return;
```

File: src/simplification/SimplifyItems.ts [hunk 2]
```diff
@@ -12,50 +13,66 @@ type Output = CommonModel | undefined;
  * @param seenSchemas already seen schemas and their corresponding output, this is to avoid circular schemas
  */
 export default function simplifyItems(schema: Schema | boolean, simplifier : Simplifier, seenSchemas: Map<any, Output> = new Map()) : Output {
-  let tempOutput: Output;
   if (typeof schema !== 'boolean') {
-    if (seenSchemas.has(schema)) return seenSchemas.get(schema)!;
+    if (seenSchemas.has(schema)) return seenSchemas.get(schema);
+    const tempOutput: Output = new CommonModel();
```
Comment (magicmatatjahu) [inline src/simplification/SimplifyItems.ts:18]: Why `tempOutput` name?

File: src/simplification/SimplifyItems.ts [hunk 3]
```diff
@@ -12,50 +13,66 @@ type Output = CommonModel | undefined;
...
-        mergeWithItem(out);
-      }
-    };
-    const handleCombinationSchemas = (schemas: (Schema | boolean)[] = []) => {
-      schemas.forEach((itemSchema) => {
-        addToItems(simplifyItems(itemSchema, simplifier, seenSchemas));
-      });
-    };
 
+    //Check if any items have been sat, if so lets merge them
```
Comment (magicmatatjahu) [inline src/simplification/SimplifyItems.ts:21]:
```suggestion
    //Check if any items have been set, if so lets merge them
```

File: src/simplification/Simplifier.ts [hunk 4]
```diff
@@ -44,9 +44,13 @@ export class Simplifier {
       this.simplifyModel(model, schema);
     }
     this.ensureModelsAreSplit(model);
+    //Ensure current model is not part of the iterated list since we could have circular schemas
+    if (this.iteratedModels[`${model.$id}`] !== undefined) {
+      delete this.iteratedModels[`${model.$id}`];
+    }
```
Comment (magicmatatjahu) [inline src/simplification/Simplifier.ts:50]: Can I trust you? With next PRs, for me logic is more and more magic 😄

**Review by magicmatatjahu — CHANGES_REQUESTED** [review]
Good job! I left two suggestions and two questions :)

File: src/simplification/SimplifyItems.ts (hunk 2 above)
Comment (jonaslagoni) [reply src/simplification/SimplifyItems.ts:18]: 🤷 you don't like it?  😄

File: src/simplification/SimplifyItems.ts (hunk 1 above)
Comment (jonaslagoni) [reply src/simplification/SimplifyItems.ts:56]: `schema` can be undefined :smile:

Comment (magicmatatjahu) [reply src/simplification/SimplifyItems.ts:56]: but then you have also false for undefined :P

File: src/simplification/Simplifier.ts (hunk 4 above)
Comment (jonaslagoni) [reply src/simplification/Simplifier.ts:50]: Never trust anyone 😈 

The problem is at line 51 we create a copy of all the values of the `iteratedModels` and if we have a circular schema that we would return the same model twice `return [model1, ...[model2, model1]];`

Unfortunately, I don't see any other way around this based on the nature of JSON Schema 😟

File: src/simplification/SimplifyItems.ts (hunk 2 above)
Comment (magicmatatjahu) [reply src/simplification/SimplifyItems.ts:18]: `temp` prefix is always used in small function and not for returned values, here you return this var. If you have no idea for a different name, leave it :)

File: src/simplification/SimplifyItems.ts (hunk 1 above)
Comment (jonaslagoni) [reply src/simplification/SimplifyItems.ts:56]: AHH, magic 🔥

File: src/simplification/SimplifyItems.ts (hunk 2 above)
Comment (jonaslagoni) [reply src/simplification/SimplifyItems.ts:18]: True, I will change it.

File: src/simplification/Simplifier.ts (hunk 4 above)
Comment (magicmatatjahu) [reply src/simplification/Simplifier.ts:52]: I understand the logic, but it seems more and more complex and I see that it (probably) will cause a lot of errors in the future. We must to think about another way to handle scheme processing.

Comment (jonaslagoni) [reply src/simplification/Simplifier.ts:52]: The only other way to do this is by not using recursion and that would be a rewrite from the ground up and basically reengineered 😄

Comment (magicmatatjahu) [reply src/simplification/Simplifier.ts:52]: Yeah, by `We must to think about another way to handle scheme processing` I mean reengineered. What we have learned is ours 😄

**Review by magicmatatjahu — APPROVED** [review]

**Pull request merged**